Write another SpinKube blogpost (#264)
spinkube/documentation: content/en/blog/news/five-new-things/index.md

---
date: 2024-11-08
title: Five New Things in SpinKube
linkTitle: Five New Things in SpinKube
description: >
  Catching up on what's new in SpinKube
author: The SpinKube Team ([@SpinKube](https://mastodon.social/@SpinKube))
resources:
  - src: "**.{png,jpg}"
    title: "Image #:counter"
---

> Since we publicly [released](/blog/2024/03/13/introducing-spinkube/) SpinKube in March we've been hard at work steadily making it better. Spin Operator [`v0.4.0`](https://github.com/spinkube/spin-operator/releases/tag/v0.4.0), ContainerD shim for Spin [`v0.17.0`](TODO), and `spin kube` plugin [`v0.3.0`](https://github.com/spinkube/spin-plugin-kube/releases/tag/v0.3.0) have all just been released. To celebrate that, here's five new things in SpinKube you should know about.

Reviewer: ```suggestion
Since we publicly [released](/blog/2024/03/13/introducing-spinkube/) SpinKube in March we've been hard at work steadily making it better. Spin Operator [`v0.4.0`](https://github.com/spinkube/spin-operator/releases/tag/v0.4.0), ContainerD shim for Spin [`v0.17.0`](https://github.com/spinkube/containerd-shim-spin/releases/tag/v0.17.0), and `spin kube` plugin [`v0.3.0`](https://github.com/spinkube/spin-plugin-kube/releases/tag/v0.3.0) have all just been released. To celebrate that, here's five new things in SpinKube you should know about.
```

> Since we publicly [released](/blog/2024/03/13/introducing-spinkube/) SpinKube in March we've been hard at work steadily making it better. Spin Operator [`v0.4.0`](https://github.com/spinkube/spin-operator/releases/tag/v0.4.0), ContainerD shim for Spin [`v0.17.0`](TODO), and `spin kube` plugin [`v0.3.0`](https://github.com/spinkube/spin-plugin-kube/releases/tag/v0.3.0) have all just been released. To celebrate that, here's five new things in SpinKube you should know about.

Reviewer 2: ```suggestion
Since we publicly [released](/blog/2024/03/13/introducing-spinkube/) SpinKube in March we've been hard at work steadily making it better. Spin Operator [`v0.4.0`](https://github.com/spinkube/spin-operator/releases/tag/v0.4.0), Containerd shim for Spin [`v0.17.0`](TODO), and `spin kube` plugin [`v0.3.0`](https://github.com/spinkube/spin-plugin-kube/releases/tag/v0.3.0) have all just been released. To celebrate that, here's five new things in SpinKube you should know about.
```

## Selective Deployments

> SpinKube now supports selectively deploying a subset of a Spin apps components. Here's an example of what the YAML for a selectively deployed app might look like:

Reviewer 2: I'd add a comment like: "you select which components you would like to be a part of the deployment in the newly added `components` field of the SpinApp

> SpinKube now supports selectively deploying a subset of a Spin apps components. Here's an example of what the YAML for a selectively deployed app might look like:

Reviewer: May want to add a bit of context on the sample app to make sure the selective deployment messaging hits home. 

```suggestion
SpinKube now supports selectively deploying a subset of a Spin apps components. Consider this simple example Spin application (named salutation in the [example repo](https://github.com/spinkube/spin-operator/tree/main/apps/salutations)) composed of two HTTP-triggered components: `hello` and `goodbye`. Here's an example of what the YAML for a selectively deployed app might look like:
```

```yaml
apiVersion: core.spinkube.dev/v1alpha1
kind: SpinApp
metadata:
  name: salutations
spec:
  image: "ghcr.io/spinkube/spin-operator/salutations:20241105-223428-g4da3171"
  executor: containerd-shim-spin
  replicas: 1
  components:
    - hello
```

> We're really excited about this feature because it makes developing microservices easier. Locally keep your whole app together and develop it as a monolith. When you go to production selectively split your app based on the characteristics of each component. For example you might run one component on nodes with GPUs and another on nodes that are closer to the end user.

Reviewer 2: ```suggestion
We're really excited about this feature because it makes developing microservices easier. Locally develop your application as a monolith in one code base. Then, when you go to production, you can split your app based on the characteristics of each component. For example you run your front end closer to the end user while keeping your backend colocated with your database.
```
I don't like the GPU example -- even though i know i started it -- because we don't support gpu access from spin apps

> We're really excited about this feature because it makes developing microservices easier. Locally keep your whole app together and develop it as a monolith. When you go to production selectively split your app based on the characteristics of each component. For example you might run one component on nodes with GPUs and another on nodes that are closer to the end user.

Reviewer: Monolith can sometimes have a negative connotation. Replacing with neutral term.
```suggestion
We're really excited about this feature because it makes developing microservices easier. Locally keep your whole app together and develop it as one. When you go to production selectively split your app based on the characteristics of each component. For example you might run one component on nodes with GPUs and another on nodes that are closer to the end user.
```

> If you want to learn more about how to use selective deployments in SpinKube checkout this [tutorial](TODO).

Reviewer: ```suggestion
If you want to learn more about how to use selective deployments in SpinKube checkout this [tutorial](https://www.spinkube.dev/docs/topics/selective-deployments/).
```

## OpenTelemetry Support

> Spin has [supported OpenTelemetry](https://www.fermyon.com/blog/spin-v24#experimental-support-for-opentelemetry-observability) for a while now, and it's now available in SpinKube. OpenTelemetry is an observability standard that makes understanding your applications running production much easier via traces and metrics.

Reviewer 3: Spin has _had_ OpenTelemetry support for a while now.

Reviewer 3: Link to OTel here?

To configure a `SpinApp` to send telemetry data to an OpenTelemetry collector you need to modify the `SpinAppExecutor`.

```yaml
apiVersion: core.spinkube.dev/v1alpha1
kind: SpinAppExecutor
metadata:
  name: otel-shim-executor
spec:
  createDeployment: true
  deploymentConfig:
    runtimeClassName: wasmtime-spin-v2
    installDefaultCACerts: true
    otel:
      exporter_otlp_endpoint: http://otel-collector.default.svc.cluster.local:4318
```

Now any Spin apps using this executor will send telemetry to the collector at `otel-collector.default.svc.cluster.local:4318`. For full details on how to use OpenTelemetry in SpinKube checkout this [tutorial](/docs/topics/monitoring-your-app).

![OpenTelemetry](/otel.png)

## MQTT Trigger Support

> The ContainerD Shim for Spin has added support for [MQTT triggers](https://github.com/spinkube/spin-trigger-mqtt). [MQTT](https://mqtt.org/) is a lightweight, publish-subscribe messaging protocol that enables devices to send and receive messages through a broker. It's used all over the place to enable Internet of Things (IoT) designs.

Reviewer 2: ```suggestion
The Containerd Shim for Spin has added support for [MQTT triggers](https://github.com/spinkube/spin-trigger-mqtt). [MQTT](https://mqtt.org/) is a lightweight, publish-subscribe messaging protocol that enables devices to send and receive messages through a broker. It's used all over the place to enable Internet of Things (IoT) designs.
```

If you want to learn more about how to use this new trigger checkout this [blog post](https://www.fermyon.com/blog/mqtt_trigger_spinkube) by Kate Goldenring.

## Spintainer Executor

> In SpinKube there is a concept of an executor. An executor is defined by the `SpinAppExecutor` CRD and it configures how a `SpinApp` is executed. Typically you'll want to define an executor that uses the ContainerD shim for Spin.

Reviewer: Is there another term we can use besides "executed" to describe the action? 
```suggestion
In SpinKube there is a concept of an executor. An executor is defined by the `SpinAppExecutor` CRD and it configures the runtime requirements for a `SpinApp`. Typically you'll want to define an executor that uses the ContainerD shim for Spin.
```

```yaml
apiVersion: core.spinkube.dev/v1alpha1
kind: SpinAppExecutor
metadata:
  name: containerd-shim-spin
spec:
  createDeployment: true
  deploymentConfig:
    runtimeClassName: wasmtime-spin-v2
    installDefaultCACerts: true
```

> However, it can also be useful to run your Spin application directly in a container. You might want to:

Reviewer: Tried to tie the executor concept in a bit more before we talk about the benefits of containerization. Let me know if that resonates with you. 

```suggestion
This configurable executor affords flexibility of running your Spin application in different environments without having to make any modifications to the `SpinApp`. For example, you may want to achieve the following by running your Spin app in a container so that you can:

- Use a specific version of Spin.
- Use a custom trigger or plugin.
- Workaround a lack of cluster permissions to install the shim.

> This is enabled by the new Spintainer executor.

Reviewer: ```suggestion
This is enabled by the new executor we've dubbed 'Spintainer'.
```

```yaml
apiVersion: core.spinkube.dev/v1alpha1
kind: SpinAppExecutor
metadata:
  name: spintainer
spec:
  createDeployment: true
  deploymentConfig:
    installDefaultCACerts: true
    spinImage: ghcr.io/fermyon/spin:v3.0.0
```

Learn more about the Spintainer executor [here](/docs/misc/spintainer-executor).

> ## Growing Stability

Reviewer: ```suggestion
## Gaining Stability
```

> SpinKube and its constituent sub-projects are all still in alpha as we leave room to iron out the kinks. However, SpinKube has made great leaps and strides in improving its stability and polish. In the number of releases we've cut for each sub-project we've squashed many bugs and sanded off plenty of rough corners.

Reviewer: Nit rephrasing

```suggestion
SpinKube and its constituent sub-projects are all still in alpha as we iron out the kinks. However, SpinKube has made great leaps and bounds in improving its stability and polish. In the number of releases we've cut for each sub-project we've squashed many bugs and sanded down plenty of rough edges.
```
